Libraries/D3D12RaytracingFallback/src/TreeletReorder.hlsl

```diff
@@ -56,210 +45,289 @@ AABB CombineAABB(AABB aabb0, AABB aabb1)
+                optimalCost[treeletBitmask] = CostOfRayBoxIntersection * optimalCost[treeletBitmask] + lowestCost; // TODO: Consider cost of flattening to triangle list
+                optimalPartition[treeletBitmask] = bestPartition;
             }
+        }
+
+        GroupMemoryBarrierWithGroupSync();
+    }
+}
+
+void ReformTree(in uint threadId)
+{
+    if (threadId != 0)
+    {
+        return;
+    }
+
+    // Now that a reordering has been calculated, reform the tree
+    struct PartitionEntry
+    {
+        uint Mask;
+        uint NodeIndex;
+    };
+    uint nodesAllocated = 1;
+    uint partitionStackSize = 1;
+    PartitionEntry partitionStack[FullTreeletSize];
+    partitionStack[0].Mask = FullPartitionMask;
+    partitionStack[0].NodeIndex = internalNodes[0];
+
+    while (partitionStackSize > 0)
+    {
+        PartitionEntry partition = partitionStack[partitionStackSize - 1];
+        partitionStackSize--;
+
+        PartitionEntry leftEntry;
+        leftEntry.Mask = optimalPartition[partition.Mask];
+        if (countbits(leftEntry.Mask) > 1)
+        {
+            leftEntry.NodeIndex = internalNodes[nodesAllocated++];
+            partitionStack[partitionStackSize++] = leftEntry;
+        }
+        else
+        {
+            leftEntry.NodeIndex = treeletToReorder[firstbitlow(leftEntry.Mask)];
+        }
+
+        PartitionEntry rightEntry;
+        rightEntry.Mask = partition.Mask ^ leftEntry.Mask;
+        if (countbits(rightEntry.Mask) > 1)
+        {
+            rightEntry.NodeIndex = internalNodes[nodesAllocated++];
+            partitionStack[partitionStackSize++] = rightEntry;
+        }
+        else
+        {
+            rightEntry.NodeIndex = treeletToReorder[firstbitlow(rightEntry.Mask)];
+        }
+
+        hierarchyBuffer[partition.NodeIndex].LeftChildIndex = leftEntry.NodeIndex;
+        hierarchyBuffer[partition.NodeIndex].RightChildIndex = rightEntry.NodeIndex;
+        hierarchyBuffer[leftEntry.NodeIndex].ParentIndex = partition.NodeIndex;
+        hierarchyBuffer[rightEntry.NodeIndex].ParentIndex = partition.NodeIndex;
+    }
 
-            // Start from the back. This is optimizing since the previous traversal went from
-            // top-down, the reverse order is guaranteed to be bottom-up
-            for (int j = numInternalTreeletNodes - 1; j >= 0; j--)
+    // Start from the back. This is optimizing since the previous traversal went from
+    // top-down, the reverse order is guaranteed to be bottom-up
+    [unroll]
+    for (int j = NumInternalTreeletNodes - 1; j >= 0; j--)
+    {
+        uint internalNodeIndex = internalNodes[j];
+        AABB leftAABB = AABBBuffer[hierarchyBuffer[internalNodeIndex].LeftChildIndex];
+        AABB rightAABB = AABBBuffer[hierarchyBuffer[internalNodeIndex].RightChildIndex];
+        AABBBuffer[internalNodeIndex] = CombineAABB(leftAABB, rightAABB);
+    }
+}
+
+void MoveToParent(in uint threadId)
```

Comment:
Worth specifying groupThreadID so the read can quickly know if this is being done once per group or one global thread